services/notifier/notifier.go

```diff
@@ -0,0 +1,588 @@
+
+	err := n.repo.resetForWorkspace(ctx, n.workspaceIdentifier)
+	if err != nil {
+		return fmt.Errorf("could not reset notifier for workspace: %s: %w", n.workspaceIdentifier, err)
+	}
+	return nil
+}
+
+func (n *Notifier) GetDBHandle() *sql.DB {
+	return n.db.DB
+}
+
+// Publish inserts the payloads into the database and returns a channel of type PublishResponse
+func (n *Notifier) Publish(
+	ctx context.Context,
+	publishRequest *PublishRequest,
+) (<-chan *PublishResponse, error) {
+	publishStartTime := n.now()
+
+	batchID := n.batchIDGenerator().String()
+
+	if err := n.repo.insert(ctx, publishRequest, n.workspaceIdentifier, batchID); err != nil {
+		return nil, fmt.Errorf("inserting jobs: %w", err)
+	}
+
+	n.logger.Infof("Inserted %d records into %s for batch: %s", len(publishRequest.Payloads), queueName, batchID)
+
+	n.stats.insertRecords.Count(len(publishRequest.Payloads))
+
+	defer func() {
+		n.stats.publishTime.Since(publishStartTime)
+		n.stats.publish.Increment()
+	}()
+
+	return n.trackBatch(ctx, batchID), nil
+}
+
+// trackBatch tracks the batch and returns a channel of type PublishResponse
+func (n *Notifier) trackBatch(
+	ctx context.Context,
+	batchID string,
+) <-chan *PublishResponse {
+	publishResCh := make(chan *PublishResponse)
+
+	n.background.group.Go(func() error {
+		defer close(publishResCh)
+
+		onUpdate := func(response *PublishResponse) {
+			select {
+			case <-ctx.Done():
+				return
+			case <-n.background.groupCtx.Done():
+				return
+			case publishResCh <- response:
+			}
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-n.background.groupCtx.Done():
+				return nil
+			case <-time.After(n.config.trackBatchInterval):
+			}
+
+			count, err := n.repo.pendingByBatchID(ctx, batchID)
+			if err != nil {
+				onUpdate(&PublishResponse{
+					Err: fmt.Errorf("could not get pending count for batch: %s: %w", batchID, err),
+				})
+				return nil
+			} else if count != 0 {
+				continue
+			}
+
+			jobs, err := n.repo.getByBatchID(ctx, batchID)
+			if err != nil {
+				onUpdate(&PublishResponse{
+					Err: fmt.Errorf("could not get jobs for batch: %s: %w", batchID, err),
+				})
+				return nil
+			}
+
+			err = n.repo.deleteByBatchID(ctx, batchID)
+			if err != nil {
+				onUpdate(&PublishResponse{
+					Err: fmt.Errorf("could not delete jobs for batch: %s: %w", batchID, err),
+				})
+				return nil
+			}
+
+			n.logger.Infof("Completed processing all files in batch: %s", batchID)
+
+			onUpdate(&PublishResponse{
+				Jobs: jobs,
+			})
+			return nil
+		}
+	})
+	return publishResCh
+}
+
+// Subscribe returns a channel of type Job
+func (n *Notifier) Subscribe(
+	ctx context.Context,
+	workerId string,
+	bufferSize int,
+) <-chan *ClaimJob {
+	jobsCh := make(chan *ClaimJob, bufferSize)
+
+	nextPollInterval := func(pollSleep time.Duration) time.Duration {
+		pollSleep = 2*pollSleep + time.Duration(rand.Intn(100))*time.Millisecond
```

Comment:
I didn't realize that since go 1.20 we didn't have to `Seed` anymore. Just pointing out in case somebody else missed it.

Reply:
Good point. Using `rand.New(rand.NewSource(time.Now().UnixNano()))`